### ISSUE TITLE
tf.GradientTape() can't train custom subclassing model.

### ISSUE THREAD

**Issue by jyh2378** [tensorflow/tensorflow#33205]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Window 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.0
- Python version: 3.7.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0, 7.6
- GPU model and memory: 8G

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I tried to make my custom model inherited from tf.keras.Model and I want to training this model by using tf.GradienTape(). 

But it can't train my model. Loss keeps same values for training and isn't reduced.
However, model.complie() and model.fit() with same criterion and optimizer work well for my custom model.
I don't understand why does this situation happen. Which part is problem?

**Code to reproduce the issue**
```
import tensorflow as tf
cfg_imagenet = [64, (128,2), 128, (256,2), 256, (512,2), 512, 512, 512, 512, 512, (1024,2), 1024]
cfg_cifar = [64, 128, 128, 256, 256, 512, 512, 512, 512, 512, 512, 1024, 1024]

class MobileNet(tf.keras.Model):
    def __init__(self, data_format='channels_last'):
        super(MobileNet, self).__init__()
        self.channel_axis = 1 if data_format == 'channels_last' else -1
        self.initializer = tf.keras.initializers.he_normal()

        self.features = self._make_layers()
        self.classifier = tf.keras.layers.Dense(10, kernel_initializer='he_normal')

    def _make_layers(self):
        layers = []
        layers.append(tf.keras.layers.Conv2D(32, kernel_size=3, strides=2, padding='same', name='conv1',
                      kernel_initializer=self.initializer))
        layers.append(tf.keras.layers.BatchNormalization(axis=self.channel_axis, name='conv1_bn'))
        layers.append(tf.keras.layers.ReLU(name='conv1_relu'))
        for idx, x in enumerate(cfg_imagenet):
            i = idx + 1
            filters = x if isinstance(x, int) else x[0]
            strides = 1 if isinstance(x, int) else x[1]
            layers.append(tf.keras.layers.DepthwiseConv2D((3, 3), strides, padding='same', kernel_initializer=self.initializer, name='block{}_dw'.format(i)))
            layers.append(tf.keras.layers.BatchNormalization(axis=self.channel_axis, name='block{}_bn1'.format(i)))
            layers.append((tf.keras.layers.ReLU(name='block{}_relu1'.format(i))))
            layers.append(tf.keras.layers.Conv2D(filters, 1, padding='same', activation='relu', kernel_initializer=self.initializer, name='block{}_pw'.format(i)))
            layers.append(tf.keras.layers.BatchNormalization(axis=self.channel_axis, name='block{}_bn2'.format(i)))
            layers.append((tf.keras.layers.ReLU(name='block{}_relu2'.format(i))))
        layers.append(tf.keras.layers.GlobalAveragePooling2D())
        return layers

    def call(self, input_tensor):
        x = input_tensor
        for layer in self.features:
            x = layer(x)
        x = self.classifier(x)
        return x

criterion = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)

def loss_func(model, x, y):
    y_ = model(x)
    return criterion(y, y_)

def grad(model, inputs, targets):
    with tf.GradientTape() as tape:
        loss_value = loss_func(model, inputs, targets)
    return loss_value, tape.gradient(loss_value, model.trainable_variables)

train_dataset = tf.data.Dataset.from_tensor_slices((train_data, train_labels))
test_dataset = tf.data.Dataset.from_tensor_slices((test_data, test_labels))
for epoch in range(10):
    train_acc = tf.keras.metrics.SparseCategoricalAccuracy(name='train_acc')
    for x, y in train_dataset:
        loss_value, grads = grad(model, x, y)
        optimizer.apply_gradients(grads_and_vars=zip(grads, model.trainable_variables))
        train_acc(y, model(x))
        print(loss_value)
        print(train_acc.result())
    print(epoch, "complete")
```

**Comment by oanush** [comment]
@jyh2378 ,
When i tried running the given code, i got the output , kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/46ccc51b036859ff3fc97cb5d120380f/33205.ipynb) of colab. Can you provide complete code to reproduce the issue mentioned here?Thanks!

**Comment by jyh2378** [comment]
@oanush 
I've got the answer of this problem. I didn't give the the parameter "training=True" to the BatchNorm layers. So, My model can't be reduced the loss. When I fix this part, model can be trained well.

**Comment by jyh2378** [comment]
@oanush 
I've got the answer of this problem. I didn't give the parameter "training=True" to BatchNorm layers.
So the model trained except parameter of BatchNorm layers. When I fixed this part, the model was trained well.
